magenta-lib/src/main/scala/magenta/deployment_type/AmiCloudFormationParameter.scala

```diff
@@ -31,21 +40,35 @@ object AmiCloudFormationParameter extends DeploymentType {
     case "update" => pkg => (reporter, lookup, parameters, stack) => {
       implicit val keyRing = lookup.keyRing(parameters.stage, pkg.apps.toSet, stack)
 
-      val stackName = stack.nameOption.filter(_ => prependStackToCloudFormationStackName(pkg))
-      val stageName = Some(parameters.stage.name).filter(_ => appendStageToCloudFormationStackName(pkg))
-      val cloudFormationStackNameParts = Seq(stackName, Some(cloudFormationStackName(pkg)), stageName).flatten
-      val fullCloudFormationStackName = cloudFormationStackNameParts.mkString("-")
+      // Based on the supplied params, decide whether to look up the CF stack by name or by tags
+      val cloudFormationStackLookupStrategy = {
+        val tagsMap = cloudFormationStackTags(pkg)
+        if (tagsMap.nonEmpty) {
+          val stackAndStage = Map("Stage" -> parameters.stage.name) ++ stack.nameOption.map("Stack" -> _)
+          val withStackAndStageTags = stackAndStage.foldLeft[Map[String, String]](tagsMap){
+            // Don't overwrite any Stack or Stage tags that the user has specified
```

Comment:
Could you achieve the same thing by starting with `stackAndStage` and then just adding `tagsMap`?